saleor/graphql/order/mutations/draft_orders.py

```diff
@@ -81,16 +87,59 @@ class Meta:
         error_type_class = OrderError
         error_type_field = "order_errors"
 
+    @classmethod
+    def validate_product_is_published_in_channel(
+        cls, info, instance, variants, channel_id
+    ):
+        channel = getattr(instance.channel, "id", None)
+        if not channel and channel_id:
+            channel = cls.get_node_or_error(info, channel_id, only_type=Channel).id
+
+        if channel:
+            variant_ids = [variant.id for variant in variants]
+            unpublished_products = ProductChannelListing.objects.filter(
+                Q(product__variants__id__in=variant_ids),
+                Q(channel__id=channel),
+                Q(is_published=False),
+            )
+            if unpublished_products:
+                raise ValidationError(
+                    {
+                        "lines": ValidationError(
+                            "Can't add unpublished products.",
```

Comment:
I would put it more explicitly. We need to make sure that our error messages precisely inform the user what happened: "Can't add products that are not published in the channel associated with this order."

Reply:
It would be really helpful if we returned IDs of variant IDs from the input that cannot be added. We can extend the `OrderError` type with `variants: [ID]` field and pass the IDs using `params` argument that is available in our validation errors.